File: Backend/src/Data/Schemas/UI/UIState.cs

using MonopolyClone.Common;
using MonopolyClone.Game;
namespace MonopolyClone.InterfaceState;

[Serializable]
public class UIState
{
    public int[] displayDices { get; set; } = new int[2] { 3, 3 };

    public bool hasPurchasedUpgrade { get; set; }

    public TurnPhase turnPhase { get; set; }

    public UIPropertyToBuy? propertyToBuy { get; set; }

    public EffectToApply? effectToAcknowledge { get; set; }

    public Auction? currentAuction { get; set; }

    public TradeState? currentTrade { get; set; }

    /// <summary>
    /// List of boolean that displays how whether the current player is connected or not.
    /// This matches exactly to the indexes of the game turn
    /// </summary>
    /// <value></value>
    public bool[] connectedUpkeep { get; set; } = new bool[0];
};

File: Frontend/src/gameState/uiState.ts
import { GameEffect } from "./effectsSchemas";
import { PropertyDeed } from "./gameState";

export const initialUI: UIState = {
  displayDices: [3, 3],
  turnPhase: 0,
  hasPurchasedUpgrade: false,
  connectedUpkeep: [],
}

export enum TurnPhase {
  Standby,
  Choiceby,
  Auctionby
};

type UIPropertyToBuy = {
  propertyID: number,
  price: number,
}

export type EffectToAcknowledge = {
  description: string,
  effect: GameEffect
};

type Bid = {
  bidder: string,
  bidAmount: number,
}

export type Auction = {
  bids: Bid[],
  topBid: number,
  currentAuctionDeadline: number,
  auctionedProperty: number,
}

export type TradeOffer = {
  properties: PropertyDeed[],
  money: number,
}

export type TradeState = {
  tradeInitiator: string,
  tradeTarget: string,

  initiatorConsent: boolean,
  targetConsent: boolean,

  initiatorOffer: TradeOffer,
  targetOffer: TradeOffer,
}


export type UIState = {
  displayDices: [number, number];
  turnPhase: TurnPhase
  propertyToBuy?: UIPropertyToBuy | null,
  effectToAcknowledge?: EffectToAcknowledge | null,
  currentAuction?: Auction | null,
  currentTrade?: TradeState | null,

  connectedUpkeep: boolean[], // whether player is connected or not
  hasPurchasedUpgrade: boolean,
};